nucliadb/src/nucliadb/common/cluster/rebalance.py

```diff
@@ -18,162 +18,444 @@
-        resource_id = result.uuid
-        try:
-            async with (
-                datamanagers.with_transaction() as txn,
-                locking.distributed_lock(
-                    locking.RESOURCE_INDEX_LOCK.format(kbid=kbid, resource_id=resource_id)
-                ),
-            ):
-                found_shard_id = await datamanagers.resources.get_resource_shard_id(
-                    txn, kbid=kbid, rid=resource_id, for_update=True
+            target_shard_capacity = max(
+                0, (settings.max_shard_paragraphs * 0.9) - target_shard.paragraphs
+            )
+            paragraphs_to_move = min(
+                shard_paragraphs - settings.max_shard_paragraphs, target_shard_capacity
+            )
+            moved_paragraphs = await self.move_paragraphs(
+                from_shard=shard, to_shard=target_shard, max_paragraphs=int(paragraphs_to_move)
+            )
+            # Update shard paragraph counts
+            shard.paragraphs -= moved_paragraphs
+            target_shard.paragraphs += moved_paragraphs
+            shards.sort(key=lambda x: x.paragraphs)
+
+            await self.wait_for_indexing()
+
+    async def merge_shard(self, shard: RebalanceShard, shards: list[RebalanceShard]):
+        logger.info(
+            "Merging shard",
+            extra={
+                "kbid": self.kbid,
+                "shard": shard.to_dict(),
+            },
+        )
+        empty_shard = False
+
+        for _ in range(MAX_MOVES_PER_SHARD):
+            resources_count = len(self.index.get(shard.id, []))
+            if resources_count == 0:
+                logger.info(
+                    "Shard is now empty",
+                    extra={
+                        "kbid": self.kbid,
+                        "shard": shard.to_dict(),
+                    },
                 )
-                if found_shard_id is None:
-                    # resource deleted
-                    continue
-                if found_shard_id != from_shard_id:
-                    # resource could have already been moved
-                    continue
-
-                await datamanagers.resources.set_resource_shard_id(
-                    txn, kbid=kbid, rid=resource_id, shard=to_shard_id
+                empty_shard = True
+                break
+
+            logger.info(
+                "Shard not yet empty",
+                extra={
+                    "kbid": self.kbid,
+                    "shard": shard.to_dict(),
+                    "remaining": resources_count,
+                },
+            )
+
+            try:
+                # Get the biggest shard with enough capacity
+                target_shard: RebalanceShard = next(
+                    reversed(
+                        [
+                            s
+                            for s in shards
+                            if s.id != shard.id and s.paragraphs < (settings.max_shard_paragraphs * 0.9)
```

Comment:
I think the active shard should be excluded as a possible target.

Reply:
The thing here is that if we exclude the active shard, we can end up in this situation:

- Let's say we have 1 shard only and it gets over the max (Shard A)
- The shard creator will create a new one (Shard B). 
- Then the rebalancer will kick in and create a third one (Shard C) where the excess of Shard A will be moved to.
- We end up with 3 shards where we could have only 2.

Maybe that's acceptable and we avoid potential problems with ingestion on the active shard?